src/main/java/com/marklogic/spark/reader/file/ZipFileReader.java

```diff
@@ -0,0 +1,94 @@
+package com.marklogic.spark.reader.file;
+
+import com.marklogic.spark.ConnectorException;
+import org.apache.hadoop.fs.Path;
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.apache.spark.sql.catalyst.expressions.GenericInternalRow;
+import org.apache.spark.sql.connector.read.PartitionReader;
+import org.apache.spark.unsafe.types.ByteArray;
+import org.apache.spark.unsafe.types.UTF8String;
+import org.apache.spark.util.SerializableConfiguration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipInputStream;
+
+public class ZipFileReader implements PartitionReader<InternalRow> {
+
+    private static final Logger logger = LoggerFactory.getLogger(ZipFileReader.class);
+
+    private final String path;
+    private final ZipInputStream zipInputStream;
+    private ZipEntry currentZipEntry;
+
+    ZipFileReader(FilePartition partition, SerializableConfiguration hadoopConfiguration) {
+        this.path = partition.getPath();
+        try {
+            if (logger.isDebugEnabled()) {
+                logger.debug("Reading zip file {}", this.path);
+            }
+            Path hadoopPath = new Path(this.path);
+            this.zipInputStream = new ZipInputStream(hadoopPath.getFileSystem(hadoopConfiguration.value()).open(hadoopPath));
```

Comment:
This is a critical bit here too - MLCP does the same thing for resolving a string path into a local/HDFS/S3 path.